### PR TITLE
Support correlated scalar subqueries (v4)

### PR DESCRIPTION
**Pull request by kokosing**
Support correlated count scalar subqueries

A scalar subquery is a simple query which aggregates all the rows into a
single group (empty GROUP BY clause). For example:

SELECT min(a) FROM t WHERE b > 0
This commit add support for correlated count scalar subqueries which
have WHERE cluase of conjunction of equality comparison. For example:

SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
Logically such query is rewritten into left outer join where correlation
used in eqality comparisons (from WHERE cluase) are used as join
condition.

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes #5984

**Comment by kokosing** [comment]
Sorted commits.

**Comment by kokosing** [comment]
Sorted and fixed some commits.

**Comment by kokosing** [comment]
There are some travis failures, I will fix them soon.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java [hunk 1]
```diff
@@ -250,10 +250,10 @@ protected void assertTableColumnNames(String tableName, String... columnNames)
         assertEquals(actual, expected);
     }
 
-    private static void assertExceptionMessage(Exception exception, @Language("RegExp") String regex)
+    private static void assertExceptionMessage(String sql, Exception exception, @Language("RegExp") String regex)
     {
         if (!exception.getMessage().matches(regex)) {
-            fail(format("Expected exception message '%s' to match '%s'", exception.getMessage(), regex));
+            fail(format("Expected query '%s' to fail with exception message '%s' matching '%s'", sql, exception.getMessage(), regex));
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java]: What if the query contains new lines? This is going to look weird. I'd rephrase it as "Expected exception message '%s' to match '%s' for query: %s"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 2]
```diff
@@ -767,6 +767,9 @@ private Object nullInputEvaluator(final Collection<Symbol> nullSymbols, Expressi
                     if (comparison.getType() == ComparisonExpression.Type.EQUAL) {
                         Set<Symbol> symbols1 = DependencyExtractor.extractUnique(comparison.getLeft());
                         Set<Symbol> symbols2 = DependencyExtractor.extractUnique(comparison.getRight());
+                        if (symbols1.isEmpty() || symbols2.isEmpty()) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: Does this cause a bug? Can you add a test?

**Comment by martint** [comment]
I've merged some of the refactoring commits. I'm looking at the rest now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 3]
```diff
@@ -311,7 +311,7 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
             }
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: This commit (in isolation) causes some unit tests to fail. For example, incorrectly complains that this query has a correlated subquery:

```sql
SELECT o1.orderkey, COUNT(*) 
FROM orders o1 LEFT OUTER JOIN (SELECT * FROM orders LIMIT 10) o2 ON EXISTS(SELECT avg(orderkey) FROM orders) 
GROUP BY o1.orderkey
```

File: presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class LongValueCompressor
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java]: I think this is overkill. The functionality we need is specific to the operator that assigns ids, so I don't see much value in abstracting it at this point. Also, this is very performance-sensitive code, so we should be careful how it's structured.

All you need is something like this:
```java
long makeId(int stageId, int taskId, int rowId)
{
    checkArgument(stageId < xxx);
    checkArgument(taskId < yyy);
    checkArgument(rowId < zzz);

    return (((long) stagedId) << aaa) | (((long) taskId) << bbb) | (long) rowId;
}
```

Given that stageId and taskId are constant for an instantiation of the operator, you could even do:

```java
long makeId(long prefix, long rowId) 
{
    checkArgument(rowId < zzz);
    return prefix | rowId;
}
```

and pre-calculate the prefix in the constructor of the operator.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java [hunk 5]
```diff
@@ -868,6 +869,15 @@ public Void visitEnforceSingleRow(EnforceSingleRowNode node, Integer indent)
         }
 
         @Override
+        public Void visitAssignUniqueId(AssignUniqueId node, Integer indent)
+        {
+            print(indent, "- Unique => [%s]", formatOutputs(node.getOutputSymbols()));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: I'd print it as AssignUniqueId to avoid confusing with "distinct".

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import static com.facebook.presto.sql.planner.optimizations.Predicates.alwaysTrue;
+import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class PlanNodeSearcher
+{
+    public static PlanNodeSearcher searchPlanNodes(PlanNode node)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: Maybe call this `searchFrom`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: Yes, it does. And unfortunately no.

Before this change, in case of comparison like `4 = a` (where `a` comes from left side of join), this method will return true, This would mean that method is a join equality comparison. Then in `visitJoin`, lines 379:
```
 boolean alignedComparison = Iterables.all(DependencyExtractor.extractUnique(equality.getLeft()), in(node.getLeft().getOutputSymbols()));
```
will say that this is an aligned comparison (left comparison symbols come from left side of join, right symbols from right) as left symbols comparison (empty set) are within left side symbols. While it is not true, as left side of comparison does not use any join symbols and right side of comparison is using symbol from left side of join.

I cannot add simple SQL test for that as usually this constant expression get projected. However, tests for rewrite for correlated subqueries will fail without this change.

File: presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java]: I can in-line this. It was even my first approach, but I had a bug there. Writing tests for an operator in isolation is not an easy task to do. So this is why I decided to extract this. 

It is easy to do a mistake when doing bitwise operation so I wanted to be sure it is well tested, as such bugs could really difficult to fix when spotted at production (it could end up with incorrect result set).

Regarding performance, IMO it is premature-optimization, such simple abstraction should be easily optimized by HotSpot.

Your call, should I in-line this?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Fixed.

**Comment by kokosing** [comment]
@martint Comments addressed, superseded by #6035

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: done

File: presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java (hunk 4 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/LongValueCompressor.java]: > I can in-line this. It was even my first approach, but I had a bug there. Writing tests for an operator in isolation is not an easy task to do. So this is why I decided to extract this.

Fair enough, but, in that case, it would suffice to have a specialized version that knows just how to assemble an id from a stageId, taskId and rowId. My main concern is with the over-abstraction it brings -- i.e., a generic utility for packing bits in a long that's used in exactly one way and one place. It's more code to understand, maintain, etc.

> Regarding performance, IMO it is premature-optimization, such simple abstraction should be easily optimized by HotSpot.

Possibly, and it may not even matter in practice given everything else that goes on in a query execution. However, from what I see in the code, it's unlikely it will perform as well as the straightforward version.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java (hunk 2 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: Thanks for the explanation. I'm adding that to the commit message.
